sympy/concrete/products.py

```diff
@@ -279,23 +298,31 @@ def _eval_product(self, term, limits):
                 return self._eval_product(factored, (k, a, n))
 
         elif term.is_Mul:
-            exclude, include = [], []
+            without_k, with_k = term.as_coeff_Mul(deps=[k])
 
-            for t in term.args:
-                p = self._eval_product(t, (k, a, n))
+            if with_k.is_Mul:
```

Comment:
Ahh, I was afraid of this and didn't see it before. The use of `as_coeff_[mul|Mul]` depends on what you are going to do with the result. Here you are going to iterate over args so you might as well have used `as_coeff_mul` which has deps built in. OK, but is deps needed on `as_coeff_Mul`? I don't think so. That's what `as_independent` is for, I think. So instead of `foo.as_coeff_Mul(deps=[k])` could you have used `as_independent(k, as_Add=False)`?

Reply:
```python
>>> (a*b*X*Y*Z*Y).as_independent(Y)  XYZ noncommutative
(a*b*X, Y*Z*Y)
```

Reply:
The idea was to factor out all the add/mul terms that didn't involve the summation variable and simply take care of those once. The rest of the logic is the same, that if the sum cannot be computed directly, take one term and try to compute the sum for that term and the remaining terms, iterate, so I do not necessarily want to take one term each time (even though I really haven't looked at the code further down, so I do not know if there are any more clever approaches for certain combinations of terms, but I guess that there may be, at least eventually) . I could have used `as_coeff_foo`, but thought it was better to use _from_args than to create new ones (which to some extent is just a matter of definition). I'll look into as_independent and see what it does.

Reply:
That did the trick! Thanks! (It felt like there should have been one, but hard to find.)

Reply:
After a more careful reading of your comment, I realize that you are correct. I'll play around a bit more with as_coeff_mul...

Reply:
My earlier reply was for the summation approach. I had forgotten about this...